ntoskrnl/io/pnpmgr/pnpnode.c

```diff
@@ -0,0 +1,104 @@
+/*
+ * PROJECT:     ReactOS Kernel
+ * LICENSE:     GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
+ * PURPOSE:     Device node handle code
+ * COPYRIGHT:   2018 Vadim Galyant
+ */
+
+#include <ntoskrnl.h>
+
+#define NDEBUG
+#include <debug.h>
+
+extern ERESOURCE PiEngineLock;
+extern ERESOURCE PiDeviceTreeLock;
+
+
+VOID
+NTAPI
+PpDevNodeLockTree(_In_ ULONG LockLevel)
+{
+    ULONG SharedCount;
+    ULONG ix;
+
+    PAGED_CODE();
+    DPRINT("PpDevNodeLockTree: LockLevel - %X\n", LockLevel);
+
+    KeEnterCriticalRegion();
+
+    switch (LockLevel)
+    {
+        case 0:
```

Comment:
In addition it would be very helpful if the locks would all be commented, explaining what they protect and how to use them, and in places like this also explain why multiple locks are acquired like they are.